File: 06_Linked_List/05_Copy_List_With_Random_Pointer/0138-copy-list-with-random-pointer.cpp
/*
Problem: LeetCode 138 - Copy List with Random Pointer

Description:
A linked list is given such that each node contains an additional random pointer that could point to any node in the list or null.
Return a deep copy of the list.

Intuition:
To create a deep copy of the given linked list, we can use a hashmap to keep track of the mapping between the original nodes and their copies. We will traverse the original list, create a copy of each node, and store the mapping in the hashmap. Then we will traverse the list again to connect the copied nodes with their corresponding random pointers.

Approach:
1. Traverse the original list, create a copy of each node, and store the mapping in the hashmap.
2. Traverse the original list again, and for each node, connect its copy with the corresponding random pointer using the hashmap.

Time Complexity:
The time complexity is O(N), where N is the number of nodes in the linked list, as we need to traverse the list twice.

Space Complexity:
The space complexity is O(N) as we use extra space to store the hashmap, where N is the number of nodes in the linked list.
*/

class Solution {
  public:
    Node *copyRandomList(Node *head) {
        if (!head) {
            return nullptr;
        }

        unordered_map<Node *, Node *> nodeMap;
        Node *current = head;

        // Traverse the original list and create a copy of each node
        while (current) {
            nodeMap[current] = new Node(current->val);
            current = current->next;
        }

        current = head;

        // Traverse the original list again to connect the copied nodes with their random pointers
        while (current) {
            nodeMap[current]->next = nodeMap[current->next];
            nodeMap[current]->random = nodeMap[current->random];
            current = current->next;
        }

        return nodeMap[head];
    }
};

File: 06_Linked_List/05_Copy_List_With_Random_Pointer/0138-copy-list-with-random-pointer.py
"""
Problem: LeetCode 138 - Copy List with Random Pointer

Key Idea:
To create a deep copy of a linked list with random pointers, we can follow a three-step approach. First, we duplicate each node in the original list and insert the duplicates right after their corresponding original nodes. Second, we update the random pointers of the duplicate nodes to point to the correct nodes. Finally, we split the combined list into two separate lists: the original list and the duplicated list.

Time Complexity:
The time complexity of this solution is O(n), where n is the number of nodes in the linked list. We traverse the list three times, once to duplicate the nodes, once to update the random pointers, and once to split the lists.

Space Complexity:
The space complexity is O(n), as we need to create a duplicate node for each original node. Additionally, we use a dictionary to map original nodes to their corresponding duplicate nodes, which takes up O(n) extra space.
"""

# Definition for a Node.
# class Node:
#     def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
#         self.val = int(x)
#         self.next = next
#         self.random = random


class Solution:
    def copyRandomList(self, head: "Node") -> "Node":
        if not head:
            return None

        # Step 1: Duplicate nodes and insert them in between the original nodes
        current = head
        while current:
            duplicate = Node(current.val)
            duplicate.next = current.next
            current.next = duplicate
            current = duplicate.next

        # Step 2: Update random pointers for the duplicate nodes
        current = head
        while current:
            if current.random:
                current.next.random = current.random.next
            current = current.next.next

        # Step 3: Split the combined list into two separate lists
        original = head
        duplicate_head = head.next
        current = duplicate_head
        while original:
            original.next = original.next.next
            if current.next:
                current.next = current.next.next
            original = original.next
            current = current.next

        return duplicate_head
